File: c++/ex5.cpp
#include<iostream>
#include<vector>
#include<cmath>
using namespace std;
double T;
double Ts;
double r;
int time_limit;
pair<double,double> aproxCof;

double korrel(const vector<double>& temperatures,double mean_y,int t)
{
    double sum_xy=0,sum_x=0,sum_y=0;
    double mean_x=(time_limit-1)/2.0;
    for(int x=0;x<=t;x++)
    {
        sum_xy+=(x-mean_x)*(temperatures[x]-mean_y);
        sum_x+=(x-mean_x)*(x-mean_x);
        sum_y+=(temperatures[x]-mean_y)*(temperatures[x]-mean_y);
    }
    return sum_xy/ sqrt(sum_x*sum_y);
}

pair<double,double> aprox(const vector<double>& x,const vector<double>& y)
{
    int n = x.size();
    double sum_xy=0,sum_x=0,sum_y=0,sum_x2=0;
    for(int i=0;i<n;i++)
    {
        sum_xy+=x[i]*y[i];
        sum_x+=x[i];
        sum_y+=y[i];
        sum_x2+=x[i]*x[i];
    }
    double a = (n*sum_xy-sum_x*sum_y)/(n*sum_x2-sum_x*sum_x);
    double b = (sum_y-a*sum_x)/n;
    pair<double,double> approx_coef(0.0,0.0);
    approx_coef.first = a;
    approx_coef.second = b;
    return approx_coef;
}

vector<pair<double,double>> coffee(double T,double Ts,double r,double time_limit)
{
    vector<double> temperatures;
    vector<double> time;
    vector<pair<double,double>> temperatures_corr;
    for(int t=0;t<=time_limit;t++)
    {
        temperatures.push_back(Ts+(T-Ts)*exp(-r*t));
        time.push_back(t);
    }

    double mean_y=0;
    for(int i=0;i<=temperatures.size();i++) 
    {
        mean_y+=temperatures[i];
    }   
    mean_y=mean_y/temperatures.size();
    for(int t=0;t<=time_limit;t++)
    {
        temperatures_corr.push_back({temperatures[t],korrel(temperatures,mean_y,t)});
    }    
    aproxCof=aprox(time,temperatures);
    return temperatures_corr;
}

int main()
{
    cout<<"T: "<<endl;
    cin>>T;
    cout<<"Ts: "<<endl;
    cin>>Ts;
    cout<<"r: "<<endl;
    cin>>r;
    cout<<"time_limit: "<<endl;
    cin>>time_limit;
    vector<pair<double,double>> t_c = coffee(T,Ts,r,time_limit);
    cout<<"коэффициенты аппроксимирующей прямой: a =  "<< aproxCof.first<<"\t"<<"b = "<< aproxCof.second<<endl;;
    cout<<"Время"<<"\t"<<"Температура"<<"\t"<<"Коэф Корреляции"<<endl;;
    int t=0;
    for(const auto& res:t_c)
    {
        cout<<t<<"\t"<<res.first<<"\t\t"<<res.second<<endl;
        t++;
    }
}

File: python/ex5.py
import math

# Global variables
T = 0.0
Ts = 0.0
r = 0.0
time_limit = 0
aproxCof = (0.0, 0.0)

def korrel(temperatures, mean_y, t):
    sum_xy = 0
    sum_x = 0
    sum_y = 0
    mean_x = (time_limit - 1) / 2.0
    for x in range(t + 1):
        sum_xy += (x - mean_x) * (temperatures[x] - mean_y)
        sum_x += (x - mean_x) ** 2
        sum_y += (temperatures[x] - mean_y) ** 2
    return sum_xy / math.sqrt(sum_x * sum_y)

def aprox(x, y):
    n = len(x)
    sum_xy = 0
    sum_x = 0
    sum_y = 0
    sum_x2 = 0
    for i in range(n):
        sum_xy += x[i] * y[i]
        sum_x += x[i]
        sum_y += y[i]
        sum_x2 += x[i] ** 2
    a = (n * sum_xy - sum_x * sum_y) / (n * sum_x2 - sum_x * sum_x)
    b = (sum_y - a * sum_x) / n
    return a, b

def coffee(T, Ts, r, time_limit):
    temperatures = []
    time = []
    temperatures_corr = []

    for t in range(time_limit + 1):
        temperatures.append(Ts + (T - Ts) * math.exp(-r * t))
        time.append(t)

    mean_y = sum(temperatures) / len(temperatures)

    for t in range(time_limit + 1):
        temperatures_corr.append((temperatures[t], korrel(temperatures, mean_y, t)))

    global aproxCof
    aproxCof = aprox(time, temperatures)
    return temperatures_corr

def main():
    global T, Ts, r, time_limit

    T = float(input("T: "))
    Ts = float(input("Ts: "))
    r = float(input("r: "))
    time_limit = int(input("time_limit: "))

    t_c = coffee(T, Ts, r, time_limit)

    print(f"aprox coaf: a = {aproxCof[0]}\tb = {aproxCof[1]}")
    print("Время\tТемпература\t\tКоэф Корреляции")

    for t, (temperature, correlation) in enumerate(t_c):
        print(f"{t}\t{temperature}\t\t{correlation}")

if __name__ == "__main__":
    main()
